core/src/main/java/org/eclipse/hono/util/TenantObject.java

```diff
@@ -311,24 +358,32 @@ public TenantObject setAdapterConfigurations(final JsonArray configurations) {
      * Adds configuration information for a protocol adapter.
      * 
      * @param config The configuration properties to add.
-     * @throws NullPointerException if config is {@code null}.
-     * @throws IllegalArgumentException if the given configuration does not contain
-     *                a <em>type</em> name.
      * @return This tenant for command chaining.
-     * @throws NullPointerException if config is {@code null}.
+     * @throws NullPointerException if the given configuration is {@code null}.
+     * @throws IllegalArgumentException if the given configuration does not contain a <em>type</em> or
+     *                                  if any of the already existing adapters has the same <em>type</em> or
+     *                                  if error parsing the given configuration.
      */
     public TenantObject addAdapterConfiguration(final JsonObject config) {
 
         Objects.requireNonNull(config);
 
-        final Object type = config.getValue(TenantConstants.FIELD_ADAPTERS_TYPE);
-        if (String.class.isInstance(type)) {
-            if (adapterConfigurations == null) {
-                adapterConfigurations = new LinkedList<>();
+        try {
+            final Adapter adapter = config.mapTo(Adapter.class);
+
+            if (adapters == null) {
+                adapters = new LinkedList<>();
+            } else {
+                final boolean hasAdapterOfSameType = adapters.stream()
+                        .anyMatch(entry -> entry.getType().equals(adapter.getType()));
+                if (hasAdapterOfSameType) {
+                    throw new IllegalArgumentException(
+                            String.format("Already an adapter of the type [%s] exists", adapter.getType()));
+                }
             }
-            adapterConfigurations.add(config.getMap());
-        } else {
-            throw new IllegalArgumentException("adapter configuration must contain type field");
+            adapters.add(adapter);
+        } catch (final IllegalArgumentException e) {
+            throw new IllegalArgumentException("Error adding the adapter configuration", e);
```

Comment:
does this really have any value?

Reply:
`config.mapTo(Adapter.class)` in this method also throws an `IllegalArgumentException` when the object is not mappable. I wanted to give more info by saying _"Error adding the adapter configuration"_ plus the reason given by `JsonObject.mapto(...)`. I think in this case, the name of this method `addAdapterConfiguration(...)` is already self explanatory. I will remove this part.